src/libmongoc/src/mongoc/mongoc-interrupt.c

```diff
@@ -0,0 +1,315 @@
+            return false;
+         }
+      }
+   }
+   return true;
+}
+
+bool
+_mongoc_interrupt_interrupt (mongoc_interrupt_t *interrupt)
+{
+   bson_mutex_lock (&interrupt->mutex);
+   if (mongoc_socket_send (interrupt->impl.socket_pair.write, "!", 1, 0) == -1 &&
+       !MONGOC_ERRNO_IS_AGAIN (errno)) {
+      _log_errno ("interrupt send failed",
+                  mongoc_socket_errno (interrupt->impl.socket_pair.write));
+      bson_mutex_unlock (&interrupt->mutex);
+      return false;
+   }
+   bson_mutex_unlock (&interrupt->mutex);
+   return true;
+}
+
+#else
+/* Pipe implementation. */
+
+/* Set non-blocking and close on exec. */
+static bool
+_set_pipe_flags (int pipe_fd)
+{
+   int flags;
+
+   flags = fcntl (pipe_fd, F_GETFL, pipe_fd);
+
+   if (-1 == fcntl (pipe_fd, F_SETFL, (flags | O_NONBLOCK))) {
+      return false;
+   }
+
+#ifdef FD_CLOEXEC
+   flags = fcntl (pipe_fd, F_GETFD, pipe_fd);
+   if (-1 == fcntl (pipe_fd, F_SETFD, (flags | FD_CLOEXEC))) {
+      return false;
+   }
+#endif
+   return true;
+}
+
+mongoc_interrupt_t *
+_mongoc_interrupt_new (uint32_t timeout_ms)
+{
+   mongoc_interrupt_t *interrupt;
+   mongoc_socket_t *interrupt_socket = NULL;
+   bool success = false;
+
+   ENTRY;
+
+   interrupt = (mongoc_interrupt_t *) bson_malloc0 (sizeof *interrupt);
+   bson_mutex_init (&interrupt->mutex);
+
+   if (0 != pipe (interrupt->impl.pipe_fds)) {
+      _log_errno ("pipe creation failed", errno);
+      GOTO (fail);
+   }
+
+   /* Make the pipe non-blocking and close-on-exec. */
+   if (!_set_pipe_flags (interrupt->impl.pipe_fds[0]) ||
+       !_set_pipe_flags (interrupt->impl.pipe_fds[1])) {
+      _log_errno ("unable to configure pipes", errno);
+   }
+
+   /* Create an unowned socket. interrupt_socket has 0 for the pid, so it will
+    * be considered unowned. */
+   interrupt_socket = bson_malloc0 (sizeof (mongoc_socket_t));
+   interrupt_socket->sd = interrupt->impl.pipe_fds[0];
+   /* Creating the stream takes ownership of the mongoc_socket_t. */
+   interrupt->stream = mongoc_stream_socket_new (interrupt_socket);
+
+   success = true;
+fail:
+   if (!success) {
+      _mongoc_interrupt_destroy (interrupt);
+      interrupt = NULL;
+   }
+   RETURN (interrupt);
+}
+
+bool
+_mongoc_interrupt_flush (mongoc_interrupt_t *interrupt)
+{
+   char c;
+   while (true) {
+      if (read (interrupt->impl.pipe_fds[0], &c, 1) == -1) {
+         if (MONGOC_ERRNO_IS_AGAIN (errno)) {
+            /* Nothing left to read. */
+            return true;
+         } else {
+            /* Unexpected error. */
+            MONGOC_ERROR ("failed to read from pipe: %d", errno);
+            return false;
+         }
+      }
+   }
+   /* Should never be reached. */
+   BSON_ASSERT (false);
+}
+
+bool
+_mongoc_interrupt_interrupt (mongoc_interrupt_t *interrupt)
+{
+   bson_mutex_lock (&interrupt->mutex);
+   if (write (interrupt->impl.pipe_fds[1], "!", 1) == -1 &&
```

Comment:
Is the `!` arbitrary?

Reply:
Yeah, https://www.man7.org/tlpi/code/online/diff/altio/self_pipe.c.html uses `x`. I also considered 'c' for libmongoc. But I chose '!' because if this character ends up visible anywhere, something definitely went wrong!